proposals/0304-structured-concurrency.md

```diff
@@ -717,119 +765,222 @@ The isCancelled function is implemented as:
+  body: (inout TaskGroup<ChildTaskResult>) async -> GroupResult
+) async -> GroupResult { ... } 
+
+
+/// Starts a new throwing task group which provides a scope in which a dynamic 
+/// number of tasks may be spawned.
+///
+/// Tasks added to the group by `group.spawn()` will automatically be awaited on
+/// when the scope exits. If the group exits by throwing, all added tasks will
+/// be cancelled and their results discarded.
+///
+/// ### Implicit awaiting
+/// When the group returns it will implicitly await for all spawned tasks to
+/// complete. The tasks are only cancelled if `cancelAll()` was invoked before
+/// returning, the groups' task was cancelled, or the group body has thrown.
+///
+/// When results of tasks added to the group need to be collected, one can
+/// gather their results using the following pattern:
+///
+///     while let result = await try group.next() {
+///       // some accumulation logic (e.g. sum += result)
+///     }
+///
+/// It is also possible to collect results from the group by using its
+/// `AsyncSequence` conformance, which enables its use in an asynchronous for-loop,
+/// like this:
+///
+///     for try await result in group {
+///       // some accumulation logic (e.g. sum += result)
+///      }
+///
+/// ### Thrown errors
+/// When tasks are added to the group using the `group.spawn` function, they may
+/// immediately begin executing. Even if their results are not collected explicitly
+/// and such task throws, and was not yet cancelled, it may result in the `withTaskGroup`
+/// throwing.
+///
+/// ### Cancellation
+/// If the task that the group is running in is cancelled, the group becomes 
+/// cancelled and all child tasks spawned in the group are cancelled as well.
+/// 
+/// If an error is thrown out of the task group, all of its remaining tasks
+/// will be cancelled and the `withTaskGroup` call will rethrow that error.
+///
+/// Individual tasks throwing results in their corresponding `try group.next()`
+/// call throwing, giving a chance to handle individual errors or letting the
+/// error be rethrown by the group.
+///
+/// Postcondition:
+/// Once `withThrowingTaskGroup` returns it is guaranteed that the `group` is *empty*.
+///
+/// This is achieved in the following way:
+/// - if the body returns normally:
+///   - the group will await any not yet complete tasks,
+///     - if any of those tasks throws, the remaining tasks will be cancelled,
+///   - once the `withTaskGroup` returns the group is guaranteed to be empty.
+/// - if the body throws:
+///   - all tasks remaining in the group will be automatically cancelled.
+static func withThrowingTaskGroup<ChildTaskResult: Sendable, GroupResult>(
+  of childTaskResult: ChildTaskResult.Type,
+  returning groupResultType: GroupResult.Type = GroupResult.self,
+  body: (inout ThrowingTaskGroup<ChildTaskResult, Error>) async throws -> GroupResult
+) async rethrows -> GroupResult { ... } 
```

Comment:
note that implementation wise this is a bit of hell --- we have to duplicate the `TaskGroup` struct impl twice, once normal and once `Throwing` but there is no way around it.

It is the "if any closure submitted to spawn throws, then next() must be throwing" so we can't just rely on rethrows.
It is even more complex with the AsyncSequence conformance, so sadly we have to duplicate implementation a bit here.